Consider hover styles when generating InteractionRegions

etiennesegonzac opened the pull request:
#### 97ea7bcbcccbf097a83ebb9416d3c12767796384
<pre>
Consider hover styles when generating InteractionRegions
<a href="https://bugs.webkit.org/show_bug.cgi?id=255940">https://bugs.webkit.org/show_bug.cgi?id=255940</a>
&lt;rdar://107265123&gt;

Reviewed by Tim Horton.

Use the presence of `:hover` styles as a signal when generating
InteractionRegions.

* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::elementMatchesHoverRules):
Add a function to detect if an element potentially matches hover rules.
(WebCore::interactionRegionForRenderedRegion):
When an element looks like a likely InteractionRegion candidate, but
doesn&apos;t have `cursor: pointer`, check to see if it matches hover rules.

* LayoutTests/interaction-region/hover-style-expected.txt: Added.
* LayoutTests/interaction-region/hover-style.html: Added.
Add test covering this logic.

Canonical link: <a href="https://commits.webkit.org/263464@main">https://commits.webkit.org/263464@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dc21342cada2e6eb99f4f3bb12817f2cc9633d34

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4422 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4542 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4674 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5901 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4621 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4411 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4721 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4506 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4850 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4486 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4658 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3974 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5902 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2157 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3951 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/7932 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3999 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4020 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5641 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4429 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3605 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3953 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3950 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1149 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8004 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4308 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -181,7 +213,17 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
     bool isTooBigForInteraction = checkedRegionArea.value() > frameViewArea / 2;
-    if (!hasListener || !hasPointer || isTooBigForInteraction) {
+
+    bool detectedHoverRules = false;
```
maybe `has`, like the others?

etiennesegonzac replied:
trying to express that maybe the element has hover styles but we didn't check

etiennesegonzac committed:
Consider hover styles when generating InteractionRegions
https://bugs.webkit.org/show_bug.cgi?id=255940
<rdar://107265123>

Reviewed by Tim Horton.

Use the presence of `:hover` styles as a signal when generating
InteractionRegions.

* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::elementMatchesHoverRules):
Add a function to detect if an element potentially matches hover rules.
(WebCore::interactionRegionForRenderedRegion):
When an element looks like a likely InteractionRegion candidate, but
doesn't have `cursor: pointer`, check to see if it matches hover rules.

* LayoutTests/interaction-region/hover-style-expected.txt: Added.
* LayoutTests/interaction-region/hover-style.html: Added.
Add test covering this logic.

Canonical link: https://commits.webkit.org/263464@main
